File: backend/internal/domain/escalation.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type EscalationPolicy struct {
	ID             uuid.UUID
	OrganizationID uuid.UUID
	Name           string
	Description    *string
	RepeatEnabled  bool
	RepeatCount    *int // NULL = infinite
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

type EscalationRule struct {
	ID              uuid.UUID
	PolicyID        uuid.UUID
	Position        int
	EscalationDelay int // minutes
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

type EscalationTarget struct {
	ID                   uuid.UUID
	RuleID               uuid.UUID
	TargetType           EscalationTargetType
	TargetID             uuid.UUID
	NotificationChannels json.RawMessage
	CreatedAt            time.Time
}

// TargetNotificationConfig represents the notification channel override for a target
type TargetNotificationConfig struct {
	Channels []string `json:"channels"` // e.g., ["email", "slack", "sms", "webhook"]
	Urgent   bool     `json:"urgent"`   // If true, use urgent/high-priority notification
}

// ParseNotificationChannels parses the notification channels config
func (t *EscalationTarget) ParseNotificationChannels() (*TargetNotificationConfig, error) {
	if len(t.NotificationChannels) == 0 {
		return nil, nil // No override configured
	}
	var config TargetNotificationConfig
	if err := json.Unmarshal(t.NotificationChannels, &config); err != nil {
		return nil, err
	}
	return &config, nil
}

// HasNotificationOverride returns true if this target has custom notification channels
func (t *EscalationTarget) HasNotificationOverride() bool {
	return len(t.NotificationChannels) > 0
}

type AlertEscalationEvent struct {
	ID               uuid.UUID
	AlertID          uuid.UUID
	PolicyID         uuid.UUID
	RuleID           *uuid.UUID
	EventType        EscalationEventType
	CurrentLevel     int
	RepeatCount      int
	NextEscalationAt *time.Time
	CreatedAt        time.Time
}

type EscalationTargetType string

const (
	EscalationTargetTypeUser     EscalationTargetType = "user"
	EscalationTargetTypeTeam     EscalationTargetType = "team"
	EscalationTargetTypeSchedule EscalationTargetType = "schedule"
)

func (t EscalationTargetType) String() string {
	return string(t)
}

func (t EscalationTargetType) Validate() error {
	switch t {
	case EscalationTargetTypeUser, EscalationTargetTypeTeam, EscalationTargetTypeSchedule:
		return nil
	default:
		return ErrInvalidEscalationTarget
	}
}

type EscalationEventType string

const (
	EscalationEventTriggered    EscalationEventType = "triggered"
	EscalationEventAcknowledged EscalationEventType = "acknowledged"
	EscalationEventCompleted    EscalationEventType = "completed"
	EscalationEventStopped      EscalationEventType = "stopped"
)

func (t EscalationEventType) String() string {
	return string(t)
}

// EscalationPolicyWithRules includes the policy and its rules with targets
type EscalationPolicyWithRules struct {
	EscalationPolicy
	Rules []*EscalationRuleWithTargets
}

// EscalationRuleWithTargets includes the rule and its targets
type EscalationRuleWithTargets struct {
	EscalationRule
	Targets []*EscalationTarget
}

File: frontend/src/lib/types/escalation.ts
export type EscalationTargetType = 'user' | 'team' | 'schedule';

export interface EscalationPolicy {
  id: string;
  organization_id: string;
  name: string;
  description?: string;
  repeat_enabled: boolean;
  repeat_count?: number;
  created_at: string;
  updated_at: string;
}

export interface EscalationRule {
  id: string;
  policy_id: string;
  position: number;
  escalation_delay: number;
  created_at: string;
  updated_at: string;
}

export interface EscalationTarget {
  id: string;
  rule_id: string;
  target_type: EscalationTargetType;
  target_id: string;
  notification_channels?: TargetNotificationConfig;
  created_at: string;
}

// Notification channel override configuration for escalation targets
export interface TargetNotificationConfig {
  channels: string[]; // e.g., ["email", "slack", "sms", "webhook"]
  urgent?: boolean; // If true, use urgent/high-priority notification
}

export interface EscalationRuleWithTargets extends EscalationRule {
  targets: EscalationTarget[];
}

export interface EscalationPolicyWithRules extends EscalationPolicy {
  rules: EscalationRuleWithTargets[];
}

// Request types

export interface CreateEscalationPolicyRequest {
  name: string;
  description?: string;
  repeat_enabled?: boolean;
  repeat_count?: number;
}

export interface UpdateEscalationPolicyRequest {
  name?: string;
  description?: string;
  repeat_enabled?: boolean;
  repeat_count?: number;
}

export interface CreateEscalationRuleRequest {
  position: number;
  escalation_delay: number;
}

export interface UpdateEscalationRuleRequest {
  position?: number;
  escalation_delay?: number;
}

export interface AddEscalationTargetRequest {
  target_type: EscalationTargetType;
  target_id: string;
  notification_channels?: TargetNotificationConfig;
}

// Response types

export interface ListEscalationPoliciesResponse {
  policies: EscalationPolicy[];
}

export interface ListEscalationRulesResponse {
  rules: EscalationRule[];
}

export interface ListEscalationTargetsResponse {
  targets: EscalationTarget[];
}
